google/cloud/spanner/result_set_test.cc

```diff
@@ -0,0 +1,190 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/spanner/result_set.h"
+#include "google/cloud/spanner/internal/time.h"
+#include "google/cloud/spanner/timestamp.h"
+#include "google/cloud/internal/make_unique.h"
+#include <gmock/gmock.h>
+#include <chrono>
+#include <string>
+
+namespace google {
+namespace cloud {
+namespace spanner {
+inline namespace SPANNER_CLIENT_NS {
+namespace {
+
+namespace spanner_proto = ::google::spanner::v1;
+
+using ::google::cloud::internal::make_unique;
+using ::testing::Return;
+
+// NOLINTNEXTLINE(clang-analyzer-optin.cplusplus.VirtualCall)
+class MockResultSetValueSource : public internal::ResultSetValueSource {
+ public:
+  // NOLINTNEXTLINE(misc-non-private-member-variables-in-classes)
+  MOCK_METHOD0(NextValue, StatusOr<optional<Value>>());
+  // NOLINTNEXTLINE(misc-non-private-member-variables-in-classes)
+  MOCK_METHOD0(Metadata, optional<spanner_proto::ResultSetMetadata>());
+  // NOLINTNEXTLINE(misc-non-private-member-variables-in-classes)
+  MOCK_METHOD0(Stats, optional<spanner_proto::ResultSetStats>());
+};
+
+TEST(ResultSet, IterateNoRows) {
+  auto mock_source_unique = make_unique<MockResultSetValueSource>();
+  auto* mock_source = mock_source_unique.get();
+  ResultSet result_set(std::move(mock_source_unique));
+
+  EXPECT_CALL(*mock_source, NextValue()).WillOnce(Return(optional<Value>()));
```

Comment:
What about structuring this like:

```cc
auto mock_source = make_unique<MockResultSetValueSource>();
EXPECT_CALL(*mock_source, NextValue()).WillOnce(Return(optional<Value>()));

ResultSet result_set(std::move(mock_source));
...
```

That makes the separate raw pointer unnecessary.